compact.go

```diff
@@ -757,9 +757,22 @@ func (c *LeveledCompactor) populateBlock(blocks []BlockReader, meta *BlockMeta,
 		}
 
 		for i, chk := range chks {
-			if chk.MinTime < meta.MinTime || chk.MaxTime > meta.MaxTime {
+			// Re-encode head chunks that are still open (being appended to) or
+			// outside the compacted MaxTime range.
+			// The chunk.Bytes() method is not safe for open chunks hence the re-encoding.
+			// This happens when snapshotting the head block.
+			//
+			// Block time range is half-open: [meta.MinTime, meta.MaxTime) and
+			// chunks are closed hence the chk.MaxTime >= meta.MaxTime check.
+			if _, isHeadChunk := chk.Chunk.(*safeChunk); isHeadChunk && chk.MaxTime >= meta.MaxTime {
+				dranges = append(dranges, Interval{Mint: meta.MaxTime, Maxt: math.MaxInt64})
+			} else if chk.MinTime < meta.MinTime || chk.MaxTime > meta.MaxTime { // Sanity check for disk blocks.
 				return errors.Errorf("found chunk with minTime: %d maxTime: %d outside of compacted minTime: %d maxTime: %d",
 					chk.MinTime, chk.MaxTime, meta.MinTime, meta.MaxTime)
+			} else if chk.MaxTime == meta.MaxTime { // This shouldn't happen as well, but will brake many users so keeping it as a warning only.
+				level.Warn(c.logger).Log("msg", "found chunk at max time boundary",
```

Comment:
And the culprit for increase allocs is this warning. I am not sure why we need warning here, we used to allow it before. When compacting 2 blocks, it is possible to have chunk MaxTime matching the meta.MaxTime. And given how many chunks there might be in a block, this will lead to 10s of thousands (or millions?) of warnings per compaction.

Reply:
I mean it would be probably enough to just check for warnings in unit test (: